storage/innobase/handler/ha_innodb.cc

```diff
@@ -12148,18 +12153,18 @@ bool create_table_info_t::create_option_encryption_is_valid() const {
   bool table_is_keyring =
       Encryption::is_keyring(m_create_info->encrypt_type.str);
 
-  if (table_is_keyring) {
-    if (!m_allow_file_per_table) {
-      my_printf_error(ER_ILLEGAL_HA_CREATE_OPTION,
-                      "InnoDB: KEYRING requires innodb_file_per_table.",
-                      MYF(0));
-      return (false);
-    }
+  if (table_is_keyring && !m_allow_file_per_table) {
+    my_printf_error(ER_ILLEGAL_HA_CREATE_OPTION,
```

Comment:
is this artificial limitation?

Because keyring encryption works on general tablespace type like system (ibdata*) and dictionary tablespace (mysql*)

Reply:
is this code for general tablespace?

But I see in encrypt_and_grep.test that it is allowed

CREATE TABLESPACE ts_rk_encrypted ADD DATAFILE 'ts_rk_encrypted.ibd' ENCRYPTION='KEYRING' ENGINE="InnoDB";

Reply:
m_allow_file_per_table is set if innodb_file_per_table is ON, but it does not block general tablespaces from being created. This is why this works. I have ported this check from MariaDB, and I was going to get back to it as it did not make much sense to me either. I have removed this check and added test in encrypt_and_grep for general tablespace (ts_k_encrypted_2) created with innodb_file_per=OFF.